### PR TITLE
trend filtering version 2

### PR DESCRIPTION
**Pull request by KaiqianZhang**
I revised trend filtering to make it work without forming any matrix. Thanks!!

### CODE REVIEW

File: R/susie_trendfilter.R [hunk 1]
```diff
@@ -1,19 +1,34 @@
...
-#' susie_trendfilter(y, 1, L=20)
-#' susie_trenndfilter(y, 0, estimate_prior_variance = TRUE)
+#' set.seed(1)
+#' mu = c(rep(0,5),rep(1,5),rep(3,5),rep(-2,5),rep(0,30))
+#' y = mu + rnorm(50)
+#' s = susie_trendfilter(y, 0)
+#' plot(y,pch=".")
+#' lines(mu,col=1,lwd=3)
+#' lines(predict(s),col=2,lwd=2)
+#' s0 = susie_trenndfilter(y, 0, estimate_prior_variance = TRUE)
```
Comment (stephens999) [inline R/susie_trendfilter.R]: Typo!

File: R/trend_filtering_multiplication.R [hunk 2]
```diff
@@ -19,30 +19,90 @@ create_Dinv = function(order, n){
...
+}
+
+#' @title Compute colSums(X * X), where X is a standardized trend filtering matrix
+#' @param order is the order of trend filtering
+#' @param n the length of y
+#' @return an n vector
+#' @keywords internal
+compute_tf_std_d = function(order, n){
+  res = rep(n-1, n)
+  if (order==0) res[n] = 0
```
Comment (stephens999) [inline R/trend_filtering_multiplication.R]: Why is this a special case?

**Comment by stephens999** [comment]
Can you add some unit tests for the compute_tf functions? I think that may also help clarify exactly what they are doing!

**Comment by KaiqianZhang** [comment]
(1) My trend filter matrix is the inverse of D^{(k+1)}, where k is the
order of trend filtering, D^{(1)} is the {-1,1} banded matrix, and
D^{(k+1)} = D^{(k)} * D^{(1)}. Should I explicitly describe it in the
documentation?

(2) I will fix the typo!

(3) This is actually by my experiments and observation. I notice that given
X is a trend filter matrix I described above, colSums(X.scaled*X.scaled) is
always rep(n-1, n) no matter what the order is.  The exception is when
order=0, it is slightly different (last entry is 0). For example, say n =
10 (n is the number of column/row of X). When order is 0,
colSums(X.scaled*X.scaled) = [9,9,9,9,9,9,9,9,9,0]. And when order is
1,2,3,..., the result is always [9,9,9,9,9,9,9,9,9,9].

**Comment by KaiqianZhang** [comment]
Sure. I will write unit tests and better document those compute_tf
functions.

**Comment by stephens999** [comment]
That is very helpful. Yes please describe in a document. What is x.scaled?
I recall that D is not invertible, as conventionally defined, so that also
requires some more explanation.

**Comment by KaiqianZhang** [comment]
Yes. I will write up a report with math proofs on trend filtering,
including how to solve the problem that D is not invertible.

What is X.scaled?
Sorry, I was not clear about X.scaled. It is X matrix after centering and
scaling, i.e.
X.scaled = safe_colScale(X).

**Comment by gaow** [comment]
Closing this obsolete pull request. A new one is to be made soon.

**Pull request closed without merging**